src/librustc_mir/interpret/memory.rs

```diff
@@ -244,41 +243,38 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         trace!("deallocating: {}", ptr.alloc_id);
 
         if ptr.offset.bytes() != 0 {
-            return err!(DeallocateNonBasePtr);
+            throw_unsup!(DeallocateNonBasePtr)
         }
 
         let (alloc_kind, mut alloc) = match self.alloc_map.remove(&ptr.alloc_id) {
             Some(alloc) => alloc,
             None => {
                 // Deallocating static memory -- always an error
-                return match self.tcx.alloc_map.lock().get(ptr.alloc_id) {
-                    Some(GlobalAlloc::Function(..)) => err!(DeallocatedWrongMemoryKind(
+                match self.tcx.alloc_map.lock().get(ptr.alloc_id) {
```

Comment:
this is more fragile now due to the changed `return`

Reply:
I agree this is one of the cases where the `return` was nice.

What happens if you just keep it? Code like this compiles:
```rust
pub fn foo(x: Option<i32>) -> Result<i32, i32> {
    return match x {
        Some(x) => return Ok(x),
        None => return Err(42),
    }
}
```
But it produces an "unreachable" warning. :/

Reply:
One option in this case would be to use
```rust
return Err(match ... { ... => err_foo!(bar), ... }.into())
```

Reply:
This looks cool!